lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -445,6 +460,70 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
                                     SN->getValue());
           RTR = F->createTranspose("transpose", NS, LTR->getShuffle());
           changed = true;
+        } else if (isa<Constant>(node->getNthInput(ArithmeticNode::LHSIdx)) &&
```

Comment:
I need to check that the `Constant` has only one use as well, right?

Reply:
Not checking isn't a correctness issue -- you're creating a new Constant and so other users should still have access to the un-transposed version. But there is a tradeoff here -- if it's a large Constant then we probably wouldn't want to make a duplicate. If it's a small Constant then maybe the compute tradeoff is worth the extra memory. For now (and if your use case only has one user) I'd say add the check.

Reply:
This is tricky thanks to the numerical correctness tests I added (hah!). The flow of a test is now like so:

1. Build graph.
2. Make pre-optimization assertions about the graph.
3. Clone the graph.
4. Optimize the cloned version of the graph.
5. Make post-optimization assertions about the optimized, cloned version of the graph.
6. Run both the original and optimized graphs with the same input and make sure the outputs are identical.

Because the graph is cloned and then its clone is optimized, these `Constants` actually have *two* uses: one in the original graph, and another in its clone.

Reply:
Hmm, yeah this is where it would be nice to have `Module::clone()` to use instead of just `Function::clone()`. Then we could copy the Constants in addition to the Function and they'd have the same number of users.

(Another place where this is a little annoying is unit tests for training, where we create two EEs/Modules: one for training and one for inference, and then we explicitly re-create the same Function twice on each. E.g. see MLTest `trainASimpleNetwork` and many others.)